Implement reset functionality

dan-palmer opened the pull request:
First of a few PR's to introduce resetting attribute values. Breaking down #115.

This just introduces the reset macro & tests the abstract form.

Commit:
Add a collection obj for resetting form attributes

Commit:
Implement resetting attributes on the base form

Commit:
Spec applying resets only when form is valid

Commit:
fixup

Commit:
Rubocop mate

Commit:
fixup

whoojemaflip commented on app/models/forms/attribute_reset_collection.rb:
```diff
@@ -0,0 +1,23 @@
+        next if form.public_send(reset_obj.master_attribute) == reset_obj.enabled_value
+
+        reset_obj.attributes_to_reset.each do |attribute|
+          form.public_send("#{attribute}=", nil)
```
should the attribute ever be reset to !nil?

dan-palmer replied:
Good point. For text fields nil seems to be the right choice. For the radio buttons I've made the assumption based on the schema.rb, it contains defaults for the text toggle fields but not anything else that is nested.

whoojemaflip replied:
this is where its a bit weird. The forms give a default value, not the models or schema

dan-palmer replied:
I did consider that, but didn't go anywhere with it when I realised we prescribe choices for some of the fields already(e.g. arson & csra) & this just felt like an extension of that. It kind of feels like theres either a small layer missing or the form objects are a little beyond what you'd expect from a form object.

whoojemaflip replied:
do you think that fields should be given a default value in the schema then?

dan-palmer replied:
We could have it look at the schema with something like this:

``` ruby
form.model.column_defaults[attribute]
```

Applied to the reset collection thing it might look something like this:

``` ruby
class AttributeResetCollection
    ...
    def perform(form)
      @collection.each do |reset_obj|
        next if form.public_send(reset_obj.master_attribute) == reset_obj.enabled_value

        reset_obj.attributes_to_reset.each do |attribute|
          reset_value = form.model.column_defaults[attribute]
          form.public_send("#{attribute}=", reset_value)
        end
      end
    end
end
```

whoojemaflip replied:
how about the default value is passed in to the `ResetData` struct by the form itself?
`ResetData = Struct.new(:attributes_to_reset, :master_attribute, :enabled_value, :default_value)`

```
concerning :ResetAttributes do
  included do
    class << self
      def reset(attributes:, if_falsey:, enabled_value: TOGGLE_YES)
        default_value = model.column_defaults[attribute]
        resettable_attributes.add(attributes, if_falsey, enabled_value, default_value)
      end
    end
  end
end
```

That would mean the `AttributeResetCollection` didn't need to know so much about the form's API

dan-palmer replied:
+1 for DI with the defaults. Did attempt that but the `model` isn't available in the class method 😢.  Instead I moved the passing to the `perform` method. Tell me whats ya fink. No rush, can discuss on Monday - have a good weekend dude.

whoojemaflip commented on spec/forms/base/resetting_attributes_spec.rb:
```diff
@@ -0,0 +1,63 @@
+
+      reset attributes: %i[ bar ], if_falsey: :foo, enabled_value: "yes"
+
+      def self.name; "LOL"; end # form base will keel over and die without it
```
😆

Commit:
Remove knowledge of form + schema source of truth

PR Feedback

Pass the dependency to the attribute collection rather than
increase the surface area of the forms api. Nothing really needs
to know about the model and even more so what the model can do.

Benefit of relying on the schemas default values for consistency
when reset is applied.

Please forgive me for mocking the AR object in the spec.

whoojemaflip commented on app/models/forms/base.rb:
```diff
@@ -12,6 +12,32 @@ class Base < Reform::Form
+        def validate(*)
+          super.tap do |valid|
+            if valid && self.class.resettable_attributes.any?
+              self.class.resettable_attributes.perform(self, default_attribute_values)
```
`.reset_all`?

Commit:
Change action method name to reflect procedure

Commit:
Use a attr reader for the reset collection

Commit:
Assert attribute is reset to default value

For the sake of this test the hash containing the defaults
has been overridden, IRL this should be the models default
column values.

whoojemaflip commented on spec/forms/base/resetting_attributes_spec.rb:
```diff
@@ -0,0 +1,69 @@
+      property :foo, type: String, validates: { presence: true }
+      property :bar, type: String
+
+      reset attributes: %i[ bar ], if_falsey: :foo, enabled_value: "yes"
```
Any plans to create an RSpec matcher to test for conformance of specific form instances?
